### DISCUSSION TITLE
Search Directory File Preview Says Existing File Doesn't Exist

### DISCUSSION THREAD

**Discussion by user9517** [PatrickF1/fzf.fish#221] [Q&A] [answered]
Hello,

When using CTRL+ALT+F to search a directory, when my pwd is `/home/{user}/Downloads`, it works great and all file and directory previews work as expected (even using exa after I configured it) and everything is displayed corretly.

But if I `cd ..` to `/home/{user}` and run CTRL+ALT+F, now the directory/file preview is saying "{dir/file} doesn't exist" for every directory/file I try to preview. Even directories/files from `/home/{user}/Downloads` that were displayed correctly from within `/home/{user}/Downloads`.

Did I do somthing wrong? Is this a known issue? How can I fix it?

My OS is: Ubuntu 21.10

All software dependency versions are:
```
fish -v         
fish, version 3.3.1

fzf --version                  
0.29.0 (dc975e8)

fd -V           
fd 8.2.1

bat --version
bat 0.19.0 (59a8f58)
```

**Comment by PatrickF1** [comment]
Hi, this might be a bug. what's your username? Does it have a space in it?

**Reply by user9517** [reply]
No, only English letters

**Reply by PatrickF1** [reply]
Is your downloads filter a symlink?

**Reply by user9517** [reply]
No downloads folder is not a symlink

**Reply by PatrickF1** [reply]
And you're executing Ctrl+Alt+F with a blank command line?

**Reply by user9517** [reply]
Correct, but i think it's the same empty or not

**Reply by PatrickF1** [reply]
Ok I think it might be related to fd, specifically this commit https://github.com/PatrickF1/fzf.fish/commit/6b592e4140168820f5df9dd28b0a93e409e0d0c3. Can you try upgrading your fd?

**Reply by PatrickF1** [reply]
btw whether or not your command line is blank actually makes a difference: https://github.com/PatrickF1/fzf.fish#file-paths

**Reply by user9517** [reply]
> Ok I think it might be related to fd, specifically this commit https://github.com/PatrickF1/fzf.fish/commit/6b592e4140168820f5df9dd28b0a93e409e0d0c3. Can you try upgrading your fd?

Do you have a specific command you want me to run? (Not near the computer that's having the problem right now, so testing will have to wait for tomorrow)

**Reply by user9517** [reply]
> btw whether or not your command line is blank actually makes a difference: https://github.com/PatrickF1/fzf.fish#file-paths

Ok interesting, i can test non empty command line too if you'd like

**Reply by PatrickF1** [reply]
> Do you have a specific command you want me to run?

No, just whatever you need to do to upgrade fd on Ubuntu.

> i can test non empty command line too if you'd like

No it's okay. It shouldn't matter. I was telling you just as an FYI so you can utilize the full powers of this plugin.

If upgrading fd doesn't work, can you send me a screenshot please?

**Reply by user9517** [reply]
> No, just whatever you need to do to upgrade fd on Ubuntu.

I mean, do you want me to install a specific version of fdfind? Or just the latest release?

> No it's okay. It shouldn't matter. I was telling you just as an FYI so you can utilize the full powers of this plugin.

Oh ok interesting, would the search behave differently? How does it add power?

> If upgrading fd doesn't work, can you send me a screenshot please?

I'll test it first thing tomorrow morning, thank you very much for the plugin and the quick response, really appreciate it.

**Reply by PatrickF1** [reply]
> I mean, do you want me to install a specific version of fdfind? Or just the latest release?

Anything past 8.3.0.

> Oh ok interesting, would the search behave differently? How does it add power?

If you type in a directory name and keep your cursor on it, it will limit the search to that directory. e.g. when in `Users/you` and you type in `Downloads/` and keep your cursor next on that word, it will limit the search to just `Downloads`.

> thank you very much for the plugin and the quick response, really appreciate it.

Glad to help!

**Reply by user9517** [reply]
> > I mean, do you want me to install a specific version of fdfind? Or just the latest release?
> 
> Anything past 8.3.0.
> 
> > Oh ok interesting, would the search behave differently? How does it add power?
> 
> If you type in a directory name and keep your cursor on it, it will limit the search to that directory. e.g. when in `Users/you` and you type in `Downloads/` and keep your cursor next on that word, it will limit the search to just `Downloads`.
> 
> > thank you very much for the plugin and the quick response, really appreciate it.
> 
> Glad to help!

updated fd version to:
```
fd -V
fd 8.3.2
```
This did not fix the issue, so I am sending a screenshot of both the correct behaviour and the buggy behaviour
This is when pwd is `/home/{user}/Downloads` and all files and folders are displayed as expected, for example you can see that content of the file hello.txt is `abc`:
![Untitled](https://user-images.githubusercontent.com/6776227/153180818-862a06f2-e34e-492f-9a58-8317c3e87db5.png)
Now after I cd into `Telegram Desktop/` (same result as cd ..), for some reason I cannot see the content of hello.txt:
![Untitled2](https://user-images.githubusercontent.com/6776227/153182880-b1d8374f-6312-48c9-b90c-1ce1610230b2.png)

While getting these screenshots, I've noticed some OTHER very strange behaviour that might give us a clue as to what is going on.
You can see in the screenshot that I have a folder in `Telegram Desktop` called `test`, and I also have a file in `Downloads` with the same exact name. After I cd into `Telegram Desktop`, for some reason, fzf.fish displays the content of `Downloads/test` (the file) as the preview for `Downloads/Telegram Desktop/test/` (the FOLDER).

And before you ask, I also tested it with a path without any spaces too (made a folder ttt inside Downloads, made a folder test inside ttt, same behaviour)

To me, it seems fzf.fish thinks that even after the cd into an inner folder, the pwd is still Downloads, so it looks for all the files/folders in Downloads, that's why it displays the wrong preview, and also why is says "doesn't exist" about names that do not exist in Downloads... VERY STRANGE BUG indeed!

Hope this helps! let me know if you need anything else from me.

**Reply by PatrickF1** [reply]
> You can see in the screenshot that I have a folder in Telegram Desktop called test, and I also have a file in Downloads with the same exact name. After I cd into Telegram Desktop, for some reason, fzf.fish displays the content of Downloads/test (the file) as the preview for Downloads/Telegram Desktop/test/ (the FOLDER).

I don't think there are both a file and a folder called `test` in the same directory. fzf.fish lists both directories and the files in them.

Ok just to cover all our bases can you run `fisher update fzf.fish` just to make sure we don't have outdated or accidentally modified code?

**Reply by user9517** [reply]
> > You can see in the screenshot that I have a folder in Telegram Desktop called test, and I also have a file in Downloads with the same exact name. After I cd into Telegram Desktop, for some reason, fzf.fish displays the content of Downloads/test (the file) as the preview for Downloads/Telegram Desktop/test/ (the FOLDER).
> 
> I don't think there are both a file and a folder called `test` in the same directory. fzf.fish lists both directories and the files in them.
> 

They are not actually in the same dir (impossible I think), what I think is going on is, for fzf.fish, for some reason, the current directory is Downloads, and even after a cd, it still looks up all files and folders relative to Downloads, so names that exist in Downloads are resolved (incorrectly, but still) and every name that is not in Downloads, just "doesn't exist".

> Ok just to cover all our bases can you run `fisher update fzf.fish` just to make sure we don't have outdated or accidentally modified code?

just did, updated to 4.3.1, same behavior.

**Reply by PatrickF1** [reply]
> They are not actually in the same dir (impossible I think), what I think is going on is, for fzf.fish, for some reason, the current directory is Downloads, and even after a cd, it still looks up all files and folders relative to Downloads, so names that exist in Downloads are resolved (incorrectly, but still) and every name that is not in Downloads, just "doesn't exist".

I am having a hard time understanding your description. Can you, either through screenshots and arrows or by drawing out the file tree, show me the layout of the files and what you mean by the wrong directory thing?

> just did, updated to 4.3.1, same behavior.

The current version is 8.1. Did you typo? Can you try following these instructions to make sure you have the latest changes? https://github.com/PatrickF1/fzf.fish/wiki/Cookbook#how-can-i-check-which-version-of-fzffish-im-on

Sorry we haven't solved your bug yet. It sure is tricky!

**Reply by user9517** [reply]
> > They are not actually in the same dir (impossible I think), what I think is going on is, for fzf.fish, for some reason, the current directory is Downloads, and even after a cd, it still looks up all files and folders relative to Downloads, so names that exist in Downloads are resolved (incorrectly, but still) and every name that is not in Downloads, just "doesn't exist".
> 
> I am having a hard time understanding your description. Can you, either through screenshots and arrows or by drawing out the file tree, show me the layout of the files and what you mean by the wrong directory thing?
> 
This is what I get when I run exa -T in Downloads/:
![image](https://user-images.githubusercontent.com/6776227/153935090-91f1d69f-8f8a-4e07-88aa-eef9c954bcf0.png)


As you can see, there are two objects named `test` here, one is a file, in `Downloads/`, containing the string "abc3" the other, is an empty directory inside the directory `ttt`. 

When I cd into `ttt`, and press CTRL+ALT+F, what fzf.fish should be showing to me is an empty directory called "test", that is because the only object inside `ttt` is an empty directory called "test".

BUT:

When I actually go inside `ttt` and press CTRL+ALT+F, fzf.fish is showing me this:

![image](https://user-images.githubusercontent.com/6776227/153935800-4d9d277f-8d69-4f55-8aea-d4b460fa61ea.png)

And because I checked, I can tell you this "abc3" is not be accident, If I change the file called "test" that is inside "Downloads" to anything else, fzf.fish will display that new value as the preview value for the empty folder that is a totally different object, In a totally different directory.

Hope I made it clear.

> > just did, updated to 4.3.1, same behavior.
> 
> The current version is 8.1. Did you typo? Can you try following these instructions to make sure you have the latest changes? https://github.com/PatrickF1/fzf.fish/wiki/Cookbook#how-can-i-check-which-version-of-fzffish-im-on
> 

OH! I was running fisher update on fzf.fish and it said version 4.3.1 but I just realized it was the version of fisher. I looked at the link you sent and it appears I have `'ppid=PARENT'` like the latest commit of 8.1 so that doesn't seem to be the issue.

Thanks!

> Sorry we haven't solved your bug yet. It sure is tricky!

That's OK, I don't blame you, it's one of the weirdest bugs i've seen as well, and I can see you are trying really hard :), that's all that matters.

**Reply by PatrickF1** [reply]
Ahhh I get it now. Thanks for explaining!
This is very mysterious indeed. Are you, by any chance, customizing the variables used to control search directory? e.g. `fzf_dir_opts`, `fzf_preview_dir_cmd`, `fzf_preview_file_cmd `, or `fzf_fd_opts`? Now I am suspecting you might have a buggy `fzf_preview_dir_cmd` that always previews $HOME or something like that.

**Reply by user9517** [reply]
Indeed very mysterious, i will print all those out for you next week.

**Reply by user9517** [reply]
> 



> Ahhh I get it now. Thanks for explaining! This is very mysterious indeed. Are you, by any chance, customizing the variables used to control search directory? e.g. `fzf_dir_opts`, `fzf_preview_dir_cmd`, `fzf_preview_file_cmd `, or `fzf_fd_opts`? Now I am suspecting you might have a buggy `fzf_preview_dir_cmd` that always previews $HOME or something like that.

If those are env vars, I don't think I have any of them set, if it's not env vars, tell me what command to run to get their values and I'll post them here.

**Reply by PatrickF1** [reply]
just run `echo $fzf_preview_dir_cmd ` etc.

**Reply by user9517** [reply]
Ok, i looked and i have none of these env vars defined... So i guess that's not it.

**Reply by PatrickF1** [reply]
Welp sorry! I have no idea what it is then. I have done as much as I can from my side without remotely controlling your computer. If I could--and you can do this if you really want to fix this bug--is to add breakpoints to the function `_fzf_search_directory` and inspecting the variables to find where the problem originates. But I'm not going to make you go through that.

**Reply by user9517** [reply]
I'd love to do that! How do you put a break point in that function?

**Reply by PatrickF1** [reply]
`breakpoint` see https://fishshell.com/docs/current/language.html#debugging-fish-scripts.
That or just echo the variables the old fashioned way.

**Comment by interestingLSY** [comment]
I have some clue... I inserted `echo (pwd)` into `_fzf_preview_file.fish` and found out that the current working directory is always `/home/{USER}` instead of (the real) working directory.

**Reply by interestingLSY** [reply]
More: Just before `_fzf_wrapper` calls `fzf`, the CWD (current working directory, can be obtained by executing `pwd`) is correct. But when `fzf` calls `_fzf_preview_file`, the CWD always points to the home directory of the user.

**Reply by interestingLSY** [reply]
When I execute `fzf --preview="pwd"` under both bash and fish shell, the output is always the home directory of the current user.

Seems that this issue is related to `fzf`?

**Reply by interestingLSY** [reply, marked as answer]
My temporary dirty hack is to replace `--preview='_fzf_preview_file {}'` with `--preview="_fzf_preview_file $(pwd)/{}"` in `--preview='_fzf_preview_file {}'`. It works.

My environment:

```plain
OS: Ubuntu 22.04
fish, version 3.6.1
fd 8.3.1
fzf 0.29 (devel)
bat 0.19.0
```

**Reply by user9517** [reply]
> More: Just before `_fzf_wrapper` calls `fzf`, the CWD (current working directory, can be obtained by executing `pwd`) is correct. But when `fzf` calls `_fzf_preview_file`, the CWD always points to the home directory of the user.

That sounds exactly like something that would cause that problem... 

@PatrickF1 can you take a look at this and tell me if you can find the bug with this new info?

**Reply by PatrickF1** [reply]
Thanks @interestingLSY for debugging. Yup, I may need to change the preview function to `--preview="_fzf_preview_file $(pwd)/{}"`. Out of curiosity, are y'all cd'ing in your config.fish? Cause by default, I think starting a new fish shell reuses your current directory.

**Reply by user9517** [reply]
> Thanks @interestingLSY for debugging. Yup, I may need to change the preview function to `--preview="_fzf_preview_file $(pwd)/{}"`. Out of curiosity, are y'all cd'ing in your config.fish? Cause by default, I think starting a new fish shell reuses your current directory.

I did not add any cd'ing in the config.fish, there may have been some added by other addons, i couldn't tell you because the Linux machine i had when i opened this issue is no longer with me, so i don't have much to add to the conversation unfortunately, hope you fix the issue anyway!

**Reply by PatrickF1** [reply]
Btw, I realized changing it to `--preview="_fzf_preview_file $(pwd)/{}"` is a bad option because it'll make the list of results in fzf much longer, unnecessarily. So instead of `file.txt`, it'll be `/Users/user/dir/file.txt`. That's bad aesthetics and not worth it for this super rare bug, if it's even a bug. I suspect when this happens something in the config is changing directories or overriding `$PWD` because the default behavior of all shells seems to be to inherit `$PWD`. If someone insists on changing directories inside their config, they're gonna have to use @interestingLSY 's hack.

**Reply by interestingLSY** [reply]
I don't change directory inside my config. I think the problem is related to `fzf`.

On my machine, when executing `fzf --preview="pwd"`, The output is always `/home/{USER}`.

Update: I re-checked my `.config/fish/config.fish` and found out that I did change the working directory in config. After removing it everything behaves correctly. Sorry.

**Reply by interestingLSY** [reply]
Replacing `--preview='_fzf_preview_file {}'` with `--preview="_fzf_preview_file $(pwd)/{}"` not seems to be a good idea since `$(pwd)` may contain special characters like quotes or spaces.

My current workaround is to insert `export FZF_CWD=(pwd)` into `_fzf_wrapper.fish` and insert `cd $FZF_CWD` into `_fzf_preview_file.fish` and `_fzf_preview_changed_file.fish`. This won't make the list of results in fzf any longer, so it is a considerable fix.

**Reply by PatrickF1** [reply]
> On my machine, when executing fzf --preview="pwd", The output is always /home/{USER}.

What version fzf are you using?

> This won't make the list of results in fzf any longer, so it is a considerable fix.

That's a great solution.

Btw I realized I misspoke when I said 

> changing it to --preview="_fzf_preview_file $(pwd)/{}" is a bad option because it'll make the list of results in fzf much longer, unnecessarily

This is not true because the list of files have already been generated at that point, and they are already relative to the current directory. Furthermore, `_fzf_preview_file` will only [print the file paths](https://github.com/PatrickF1/fzf.fish/blob/main/functions/_fzf_preview_file.fish#L33) when the file is not a regular file or a directory. So it's not as bad as I thought. That said, I am still hesitant because Search Directory is becoming way too complicated and responsible for 80% of the reported bugs and I regret ever creating it. And as you said, "$(pwd) may contain special characters like quotes or spaces." I will not fix this potential bug unless it turns out this is a very common issue. Sorry!
